src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -693,6 +700,85 @@ protected override void SetItem(int index, string item)
             {
                 base.SetItem(index, string.IsNullOrEmpty(item) || item == "*.*" ? "*" : item);
             }
+
+            internal string[] GetFilters() => ((ImmutableStringList)Items).Items;
+
+            /// <summary>
+            /// List that maintains its underlying data in an immutable array, such that the list
+            /// will never modify an array returned from its Items property. This is to allow
+            /// the array to be enumerated safely while another thread might be concurrently mutating
+            /// the collection.
+            /// </summary>
+            private sealed class ImmutableStringList : IList<string>
+            {
+                public string[] Items = Array.Empty<string>();
+
+                public string this[int index]
+                {
+                    get
+                    {
+                        string[] items = Items;
+                        if ((uint)index >= (uint)items.Length)
+                        {
+                            throw new ArgumentOutOfRangeException(nameof(index));
+                        }
+                        return items[index];
+                    }
+                    set
+                    {
+                        string[] clone = (string[])Items.Clone();
```

Comment:
if the previous and new value of ```clone[index]``` should we skip the cloning part in that case ?

Reply:
Is that common?  My expectation is "no", in which case I'd prefer we just pay the copy costs in the rare case where they are equal, rather than always paying the string comparison costs in the case that they are.